src/tests.rs

```diff
@@ -4506,3 +4506,662 @@ fn test_key() {
 }
+
+use std::io;
+use std::io::BufRead;
+
+#[derive(Debug, PartialEq)]
+enum KdfCtrlLoc {
+    Undefined,
+    AfterFixed,
+    AfterIter,
+    BeforeFixed,
+    BeforeIter,
+    MiddleFixed,
+}
+
+#[derive(Debug)]
+struct KdfTestUnit {
+    line: usize,
+    count: usize,
+    l: usize,
+    ki: Vec<u8>,
+    iv_len: usize,
+    iv: Vec<u8>,
+    data_len: usize,
+    data: Vec<u8>,
+    data_before_len: usize,
+    data_before: Vec<u8>,
+    ko: Vec<u8>,
+}
+
+#[derive(Debug)]
+struct KdfTestSection {
+    kdf: CK_MECHANISM_TYPE,
+    prf: CK_MECHANISM_TYPE,
+    ctr_location: KdfCtrlLoc,
+    rlen: usize,
+    units: Vec<KdfTestUnit>,
+}
+
+macro_rules! parse_or_panic {
+    ($e:expr; $line:expr; $ln:expr) => {
+        match $e {
+            Ok(r) => r,
+            Err(_) => panic!("Malformed line '{}' (line {})", $line, $ln),
+        }
+    };
+}
+
+fn parse_kdf_vector(filename: &str) -> Vec<KdfTestSection> {
+    let file = ret_or_panic!(std::fs::File::open(filename));
+
+    let mut kdf: CK_MECHANISM_TYPE = CK_UNAVAILABLE_INFORMATION;
+    let mut data = Vec::<KdfTestSection>::new();
+
+    for (l, line) in io::BufReader::new(file).lines().flatten().enumerate() {
+        let ln = l + 1;
+        if line.starts_with("#") {
+            match line.as_str() {
+                "# KDF Mode Supported: Counter Mode" => {
+                    kdf = CKM_SP800_108_COUNTER_KDF
+                }
+                "# KDF Mode Supported: Feedback Mode" => {
+                    kdf = CKM_SP800_108_FEEDBACK_KDF
+                }
+                _ => (),
+            }
+            continue;
+        }
+
+        if line.len() == 0 {
+            continue;
+        }
+
+        if line.starts_with("[PRF=") {
+            /* we ignore tests for algorithms we do not care to support like Triple DES,
+             * for those we still need to parse the section, but we'll mark it as
+             * unknown and skip all units */
+            let section = KdfTestSection {
+                kdf: kdf,
+                prf: match &line[5..] {
+                    "CMAC_AES128]" => CKM_AES_CMAC,
+                    "CMAC_AES192]" => CKM_AES_CMAC,
+                    "CMAC_AES256]" => CKM_AES_CMAC,
+                    "HMAC_SHA1]" => CKM_SHA_1_HMAC,
+                    "HMAC_SHA224]" => CKM_SHA224_HMAC,
+                    "HMAC_SHA256]" => CKM_SHA256_HMAC,
+                    "HMAC_SHA384]" => CKM_SHA384_HMAC,
+                    "HMAC_SHA512]" => CKM_SHA512_HMAC,
+                    _ => CK_UNAVAILABLE_INFORMATION,
+                },
+                ctr_location: KdfCtrlLoc::Undefined,
+                rlen: 0,
+                units: Vec::with_capacity(39),
+            };
+            data.push(section);
+            continue;
+        }
+        let section = match data.last_mut() {
+            Some(s) => s,
+            None => continue,
+        };
+        if section.prf == CK_UNAVAILABLE_INFORMATION {
+            continue;
```

Comment:
just an idea, likely bad. Would it make sense to parse also these, pass them to the pkcs11 module and just expect graceful error as unsupported rather than skipping altogether?

Reply:
Not sure why that would help, these are tests that us a prf we do not parse/support.
I could have simply removed the tests from the files (like they did in NSS), but I preferred to just ignore unsupportable algorithms (these are Three DES stuff). The supportable, but unimplemented algorithm (AES_CMAC) you can see is skipped explitily later and is instead parsed.